File: 15-3sum.rs
// 3Sum
// #url: https://leetcode.com/problems/3sum/
// #medium
//
// Solution:
//   https://en.wikipedia.org/wiki/3SUM

use std::collections::HashMap;

fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
  if nums.len() < 3 {
    return vec![];
  }

  let mut sorted_nums = nums.clone();
  sorted_nums.sort();
  let mut i: usize = 0;
  let mut results = HashMap::new();

  while i < sorted_nums.len() - 2 {
    let a = sorted_nums[i];
    let mut start = i + 1;
    let mut end = sorted_nums.len() - 1;

    while start < end {
      let b = sorted_nums[start];
      let c = sorted_nums[end];

      if a + b + c == 0 {
        results.insert(format!("{}{}{}", a, b, c), vec![a, b, c]);
        start = start + 1;
        end = end - 1;
      } else if a + b + c > 0 {
        end = end - 1;
      } else {
        start = start + 1;
      }
    }
    i += 1;
  }

  results.iter().map(|(_, res)| res.clone()).collect()
}

fn main() {
  println!("{:?}", three_sum(vec![-1, 0, 1, 2, -1, -4]))
}

#[cfg(test)]
mod tests {
  // Note this useful idiom: importing names from outer (for mod tests) scope.
  use super::*;

  #[test]
  fn test_1() {
    assert_eq!(true, true);
  }
}

File: 15-3sum.js
// 3Sum
// #url: https://leetcode.com/problems/3sum/
// #medium
//
// Solution:
//   https://en.wikipedia.org/wiki/3SUM

// function treeSum(arr, target = 0) {
//   if (arr.length < 3) return [];
//   arr = arr.sort((a, b) => a - b);

//   let result = {};
//   let i = 0;
//   while (i < arr.length - 2) {
//     let a = arr[i];
//     let start = i + 1;
//     let end = arr.length - 1;

//     while (start < end) {
//       let b = arr[start];
//       let c = arr[end];

//       if (a + b + c === target) {
//         result[`${a}${b}${c}`] = [a, b, c];
//         start++;
//         end--;
//       } else if (a + b + c > target) {
//         end--;
//       } else {
//         start++;
//       }
//     }

//     i++;
//   }
//   return Object.keys(result).map(key => result[key]);
// }

function treeSum(nums, target = 0) {
  if (nums.length < 3) return [];
  let result = [];
  for (let i = 0; i < nums.length; i++) {
    let set = new Set();
    let sum = target - nums[i];
    for (let j = i + 1; j < nums.length; j++) {
      if (set.has(sum - nums[j])) {
        console.log(nums[i], nums[j], sum - nums[j]);
        let item = [nums[i], nums[j], sum - nums[j]].sort((a, b) => a - b);
        result[item.join("")] = item;
      }
      set.add(nums[j]);
    }
  }

  return Object.keys(result).map(key => result[key]);
}

console.log(treeSum([-1, 0, 1, 2, -1, -4]), [[-1, 0, 1], [-1, -1, 2]]);
